File: mobile_app/lib/features/products/model/product.dart
// Model class representing a product in the shop
class Product {
  final String id;
  final String name;
  final List<double> location;
  final String series;
  final double priceOriginal;
  final double priceUsers;
  final String expDate;
  final String category;
  final String store;
  final int storeId;
  final int quantity;
  final String photoUrl;
  Product({
    required this.id,
    required this.name,
    required this.location,
    required this.series,
    required this.priceOriginal,
    required this.priceUsers,
    required this.expDate,
    required this.category,
    required this.store,
    required this.storeId,
    required this.quantity,
    required this.photoUrl,
  });
  factory Product.fromJson(Map<String, dynamic> json) {
    final List<dynamic>? rawLocation = json['location'] as List<dynamic>?;
    final List<double> parsedLocation = [];
    if (rawLocation != null && rawLocation.length >= 2) {
      try {
        parsedLocation.add((rawLocation[0] as num).toDouble());
        parsedLocation.add((rawLocation[1] as num).toDouble());
      } catch (e) {
        print('Error parsing location coordinates: $e');
      }
    }
    if (parsedLocation.isEmpty) {
      parsedLocation.addAll([0.0, 0.0]);
    }
    return Product(
      id: json['id'] as String,
      name: json['name'] as String,
      location: parsedLocation,
      series: json['series'] as String,
      priceOriginal: (json['price_original'] as num).toDouble(),
      priceUsers: (json['price_users'] as num).toDouble(),
      expDate: json['exp_date'] as String,
      category: json['category'] as String,
      store: json['store'] as String,
      storeId: json['store_id'] as int,
      quantity: json['quantity'] as int,
      photoUrl: json['photo_url'] as String,
    );
  }
}

File: backend/classes/product.py
from classes.store import Store
from classes.database_interface import DatabaseInterface

class Product:
    def __init__(self, name, series, price_original, price_users, exp_date, EAN, category, store : Store, quantity, photo_url):
        self.name = name
        self.series = series
        self.price_original = price_original
        self.price_users = price_users
        self.exp_date = exp_date
        self.EAN = EAN
        self.category = category
        self.store = store
        self.quantity = quantity
        self.photo_url = photo_url
        self.id = f"{self.store.id}_{self.EAN}_{self.series}"

    @classmethod
    def from_database(cls, doc, database : DatabaseInterface):
        store = database.get_store(doc["store_id"])
        return cls(
            name = doc["name"],
            series = doc["series"],
            price_original = doc["price_original"],
            price_users = doc["price_users"],
            exp_date = doc["exp_date"],
            EAN = doc["EAN"],
            category = doc["category"],
            store = store,
            quantity = doc["quantity"],
            photo_url = doc["photo_url"]
        )

    def get_location_coords(self):
        return self.store.get_location().get_coords()
    
    def get_city(self):
        return self.store.get_location().get_city()
    
    def prepare_dict(self):
        return dict({
            "id" : self.id,
            "name" : self.name,
            "location" : self.get_location_coords(),
            "city" : self.get_city(),
            "series": self.series,
            "price_original": self.price_original,
            "price_users": self.price_users,
            "exp_date": self.exp_date,
            "EAN": self.EAN,
            "category": self.category,
            "store": self.store.name,
            "store_id" : self.store.id,
            "quantity": self.quantity,
            "photo_url" : self.photo_url
        })
    

def get_all_products(database : DatabaseInterface):
    product_records = database.get_all_products()
    product_objects = []
    for rec in product_records:
        product_objects.append(Product.from_database(rec, database))
    return product_objects
